# Day 1: Trebuchet?!

Here's the first puzzle:

> <p>Something is wrong with global snow production, and you've been selected to take a look. The Elves have even given you a map; on it, they've used stars to mark the top fifty locations that are likely to be having problems.</p>
> <p>You've been doing this long enough to know that to restore snow operations, you need to check all <em class="star">fifty stars</em> by December 25th.</p>
> <p>Collect stars by solving puzzles.  Two puzzles will be made available on each day in the Advent calendar; the second puzzle is unlocked when you complete the first.  Each puzzle grants <em class="star">one star</em>. Good luck!</p>
> <p>You try to ask why they can't just use a <a href="/2015/day/1">weather machine</a> ("not powerful enough") and where they're even sending you ("the sky") and why your map looks mostly blank ("you sure ask a lot of questions") <span title="My hope is that this abomination of a run-on sentence somehow conveys the chaos of being hastily loaded into a trebuchet.">and</span> hang on did you just say the sky ("of course, where do you think snow comes from") when you realize that the Elves are already loading you into a <a href="https://en.wikipedia.org/wiki/Trebuchet" target="_blank">trebuchet</a> ("please hold still, we need to strap you in").</p>
> <p>As they're making the final adjustments, they discover that their calibration document (your puzzle input) has been <em>amended</em> by a very young Elf who was apparently just excited to show off her art skills. Consequently, the Elves are having trouble reading the values on the document.</p>
> <p>The newly-improved calibration document consists of lines of text; each line originally contained a specific <em>calibration value</em> that the Elves now need to recover. On each line, the calibration value can be found by combining the <em>first digit</em> and the <em>last digit</em> (in that order) to form a single <em>two-digit number</em>.</p>
> <p>For example:</p>
> <pre><code>1abc2
> pqr3stu8vwx
> a1b2c3d4e5f
> treb7uchet
> </code></pre>

Let's use regular expressions to solve it. Note that the first cell here contains our imports, and the cell is marked as "parameters" (that's a cell tag). This will allow us to inject different `INPUT` values in to test our notebook against different conditions. Your `OUTPUT` (your solution to the puzzle) should appear in a different cell than the `parameters`-tagged cell, in order for the parameter injection to work.

In [ ]:
from re import MULTILINE, VERBOSE, compile

from advent23 import disp_name

INPUT = """\
1abc2
pqr3stu8vwx
a1b2c3d4e5f
treb7uchet
"""

## Find matches

Refer to the "RegEx Reference" sidebar menu at [regexr](https://regexr.com). Compile a regex pattern for matching the first and last digits in a line.

In [ ]:
pattern = compile(
    # Verbose: ignore implicit whitespace and allow comments in the pattern below
    # Multiline: ^ and $ match the start of a line rather than the end of the string
    flags=VERBOSE | MULTILINE,
    pattern=(
        r"""
        ^               # Start of line
        [^\d]*          # Zero or more non-digits
        (?P<first>\d)   # first digit
        .*?             # Zero or more characters, lazily
        (?P<last>\d)?   # last digit, optional
        [^\d]*          # Match zero or more non-digits
        $               # Match end of line
        """
    ),
)

matches = pattern.findall(INPUT)
disp_name("All matches:", matches)

#### All matches:

[('1', '2'), ('3', '8'), ('1', '5'), ('7', '')]

## Handle matches

The comprehension filters out non-matches.

In [ ]:
calibration_values = []
for match in [m for m in matches if m]:
    first, last = match
    last = last if last else first
    calibration_values.append(int(f"{first}{last}"))

disp_name("Calibration values:", calibration_values)

#### Calibration values:

[12, 38, 15, 77]

> <p>In this example, the calibration values of these four lines are <code>12</code>, <code>38</code>, <code>15</code>, and <code>77</code>.

## Sum the calibration values

In [ ]:
OUTPUT = sum(calibration_values)
disp_name("Sum:", OUTPUT)

#### Sum:

142

> Adding these together produces <code><em>142</em></code>.</p>
> <p>Consider your entire calibration document. <em>What is the sum of all of the calibration values?</em></p>
> </article>